### ISSUE TITLE
Add filter linter

### ISSUE THREAD

**Issue by jspenguin2017** [NanoAdblocker/NanoCore#1]
Investigate what needs to be changed in order to show the user why each filter is rejected. 
Also maybe show a warning when a duplicate filter is already present?

**Comment by jspenguin2017** [comment]
@ameshkov I'm not sure if this works in Adguard, but it's rejected by uBO (and Nano) and looks kind of weird. The scope operator doesn't point to anything, and immediately followed by `:match-css`
![image](https://user-images.githubusercontent.com/7283682/34429143-5c4ecc10-ec11-11e7-94a3-413008b0ff36.png)

**Comment by ameshkov** [comment]
@jspenguin2017 
> The scope operator doesn't point to anything, and immediately followed by :match-css. Is that intended?

Yes, this is a valid pseudo-class usage. @gorhill, looks like a bug to me.

> Also, shouldn't this be #$#?

Absolutely, thx for pointing this out!

**Comment by gorhill** [comment]
> Yes, this is a valid pseudo-class usage. @gorhill, looks like a bug to me.

You need to use a `*`: `*:matches-css(...)`. Pseudo operators are meant to mimick what could be a valid CSS selector, and without a prefix this wouldn't work even for valid CSS selectors. Example: `.headeliner > :not(...)` is not a valid selector, or assuming `:has` is implemented by browser vendors, `.headeliner > :has(...)` wouldn't be valid selector.

**Comment by ameshkov** [comment]
> Example: .headeliner > :not(...) is not a valid selector

It is a valid selector:
![](https://uploads.adguard.com/up04_pzxle_Nano_Linter_issues_mega_thread__Issue_1__NanoAdblockerNanoCore_.png)

**Comment by gorhill** [comment]
Ok, well, today-I-learned. I could have just tried it at the console... I will need to fix in uBO.

**Comment by jspenguin2017** [comment]
@ameshkov So the `*` is implicit? Then it would be a duplicate of the rule below?

**Comment by ameshkov** [comment]
> @ameshkov So the * is implicit?

Yes, it is.

> Then it would be a duplicate of the rule below?

As I understand, it's not.

In the rule below it means that there is an intermediate element between `.page-form` and `:matches-css`.

**Comment by jspenguin2017** [comment]
It looks like having `> :` is equivalent to just `:`, so `*` is not implicit, `>` here is instead ignored. 
![image](https://user-images.githubusercontent.com/7283682/34450938-ca245862-ecd5-11e7-84f6-829a25a72524.png)
Or I'm still missing something?

**Comment by jspenguin2017** [comment]
Looks like it's not quite implicit `*`, the 3 cases are different. 
![image](https://user-images.githubusercontent.com/7283682/34450961-307f1b9c-ecd6-11e7-8b5c-9652712ec835.png)

**Comment by jspenguin2017** [comment]
If I understood right: 
- `elem > :op` means apply the operator `op` to immediate children of `elem`
- `elem > * :op` means apply the operator `op` to any (grand)children of any immediate children of `elem`
- `elem :op` means apply the operator `op` to any (grand)children of `elem`

**Comment by grenzor** [comment]
Not sure if you know about these but:

1) `||example.com$important,~document` discards the filter and gives a cannot be negated error, but `document` can be negated and the filter should be valid.

2) `@||example.com` no error is given but rule is missing a @ at the start

**Comment by jspenguin2017** [comment]
@grenzor 
The liner is *always* right when it comes to showing *error*, because it uses the exact same parsing engine. If it shows error, then the filter is rejected, if it does not show error, then the filter is accepted, although how will it behave is another question. 

I do agree warnings should be shown for some of these cases. I'll work on that later.

**Comment by grenzor** [comment]
>The error in `||example.com^,domain=example2.com` and `||example.com^$third-party,$image` are pretty obvious from syntax highlighting

They may be easy to spot if you have a few filters, but for someone with 100s/1000s of their own filters it is not easy at all. It's easier for the user to scroll down and just look at the sidebar for an icon of error or warning than manually checking each line's color syntaxing.

**Comment by jspenguin2017** [comment]
@grenzor 
I think you should be writing one filter at a time, and not 1000 and test together. Beside, even if the filter doesn't contain obvious mistake, it doesn't mean it'll work.

**Comment by grenzor** [comment]
I agree, but I guess my use case was someone with already written filters on uBO who then imported them into Nano and tried use its linter to check the validity of these filters.

**Comment by jspenguin2017** [comment]
@grenzor 
You should have still validated each filter when you wrote them to make sure they actually works, and not write them and hope it's good.

**Comment by jspenguin2017** [comment]
Is this clear enough? Or I should still dispatch a linter warning? 
![image](https://user-images.githubusercontent.com/7283682/36880474-1b7fe9b6-1d86-11e8-9fb6-9f57c9213f5f.png)

**Comment by jspenguin2017** [comment]
@ameshkov FYI, syntax errors in the spyware filter:
```
gopro.com#$#body[class^="smartling-] { padding-top: 60px!important; }
xtreme.ws##.topBorder > table[width="100%"][height="205"]body > #top
```

First one missing other double-quote, second one is invalid selector.

**Comment by ameshkov** [comment]
@jspenguin2017 thank you, we'll fix them soom!

**Comment by jspenguin2017** [comment]
- ~~`@@||example.com^generichide` - Forgot `$`~~ Pretty obvious from highlighting
It is quite rare that `^` is used just to match `/`, 9 out of 10 times it's a mistake. 
Detection logic: If a filter has one `^` and no `$`, and `^` is not at the very end, then insert `$` right after `^`, if the filter is still (or still seems) valid, then dispatch a warning. 
- ~~`!#incluuuude` - Misspelled preprocessor (need to investigate how preprocessor is handled for `My filters`)~~ Preprocessors will be highlighted as keywords
Preprocessor are like comments, but has special meaning when spelled correctly. 
Detection logic: If a line starts with `! #`, then dispatch a warning (extra space). Otherwise, if a line starts with `!#` but is not a valid preprocessor directive, dispatch a warning.

**Comment by jspenguin2017** [comment]
Need to investigate how the redirect feature works with RegExp rules. With recent change (https://github.com/uBlockOrigin/uBlock-issues/issues/977), the redirect portion of RegExp rules is no longer rejected.
